blocks/monitoring/__init__.py

```diff
@@ -0,0 +1,193 @@
+"""This part of the framework helps you monitor your model during training.
+
+:class:`MonitorChannels` describe quantities that should be monitored. They
+can either be Theano variables (e.g. the objective function, the weight
+norms, etc.) or they can be callable functions (e.g. if you want to use
+NumPy to calculate eigenvalues, or sample text from your language model).
+
+.. todo::
+
+   Monitored Theano variables might depend on inputs beyond those given to
+   the model. There should be a way of providing these to the monitoring
+   channels, collecting them, and compiling them as part of the Theano
+   function that performs monitoring.
+
+"""
+import logging
+from abc import ABCMeta, abstractmethod
+
+from blocks.utils import shared_expression, update_instance
+
+logger = logging.getLogger(__name__)
+
+
+class MonitorChannel(object):
```

Comment:
I don't understand what you mean with having many different sources with the same name? If you mean across different datasets, then it is true that the source is always called `features`. However, I don't think that is much of a problem. We can't guarantee global uniqueness anyway, so if people want to combine multiple datasets they might as well prefix the dataset name in front of the source.

The ability to ask a dataset for specific sources only might be worth considering. It would be a shame if the user is forced to read large amounts of data from disk just to reject them afterwards (e.g. if a dataset consists of images, tags, and a category, and we only want the tags and category).